subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/projectmodule/DefaultProjectDependencyPublicationResolver.java

```diff
@@ -93,10 +124,41 @@ private <T> T resolve(Class<T> coordsType, Path identityPath, ProjectInternal pr
         }
 
         if (topLevelWithComponent.size() == 1) {
-            return topLevelWithComponent.iterator().next().getCoordinates(coordsType);
+            SoftwareComponentInternal singleComponent = topLevelWithComponent.iterator().next().getComponent().get();
+            Map<SoftwareComponent, T> componentCoordinates = getComponentCoordinates(coordsType, publications.keySet());
+            return new MultiCoordinateVariantResolver<>(singleComponent, identityPath, componentCoordinates);
         }
 
         // See if all entry points have the same identifier
+        return VariantCoordinateResolver.of(getCommonCoordinates(project, coordsType, topLevel));
+    }
+
+    /**
+     * Get the coordinates of a project that has no publications.
+     */
+    private static <T> T getImplicitCoordinates(Class<T> coordsType, Project project) {
+        // Project has no publications: simply use the project name in place of the dependency name
+        if (coordsType.isAssignableFrom(ModuleVersionIdentifier.class)) {
+
+            // TODO: Deprecate this behavior
+//        String message = "Cannot publish dependency on " + project.getDisplayName() + " since it does not declare any publications. " +
+//            "Publishing a component that depends on another project without publications";
+//        DeprecationLogger.deprecate(message)
+//            .withAdvice("Ensure " + project.getDisplayName() + " declares at least one publication.")
+//            .willBecomeAnErrorInGradle9()
+//            .withUpgradeGuideSection(8, "publishing_dependency_on_unpublished_project")
+//            .nagUser();
+
+            return coordsType.cast(DefaultModuleVersionIdentifier.newId(project.getGroup().toString(), project.getName(), project.getVersion().toString()));
```

Comment:
This is existing code. I didn't write any of this.

However, I'm not sure what other option you are suggesting. I am open to ideas but I don't see another way to get this to compile. The method passes in a `Class<T>` and returns a `T`. We need an instance of `T` to return. So, this code checks that the passed-in class is a module version ID and if so returns a module version ID. 

The cast is unavoidable since we need a `T` to return. If there is another form you can think of that does the same thing that also compiles, we can move to that

Reply:
What I meant was to change the `isAssignableFrom` to check `DefaultModuleVersionIdentifier`, since that is the type we're casting to, instead of merely `ModuleVersionIdentifier`, which may at some point get multiple implementations.  If that happened, we'd fail with a casting exception on 153, instead of what looks to be the desired `UOE` on 156.

Reply:
I see what you're saying now. I changed this to check that the class is equal to a `ModuleVersionIdentifier` instead of assignable from a `ModuleVersionIdentifier`. 

We can't check for `DefaultModuleVersionIdentifier` since the value of `coordsType` is `ModuleVersionIdentifier` and that would never match.